src/test/path-mapping.spec.ts

```diff
@@ -0,0 +1,129 @@
+import { join } from 'path';
+
+import { createExec } from './exec-helpers';
+import {
+  CMD_ESM_LOADER_WITHOUT_PROJECT,
+  CMD_TS_NODE_WITHOUT_PROJECT_FLAG,
+  nodeSupportsEsmHooks,
+  TEST_DIR,
+} from './helpers';
+import { test } from './testlib';
+
+const execBuilder = (
+  command: string,
+  moduleDir: string,
+  tsConfig: string = 'tsconfig.json'
+) => {
+  const partialExec = createExec({
+    cwd: join(TEST_DIR, moduleDir),
+    env: { ...process.env, TS_NODE_PROJECT: tsConfig },
+  });
+
+  return (file = 'index.ts', throwErrors = true) =>
+    partialExec(`${command} ${file}`).then((res) => {
+      if (throwErrors && res.err) throw res.err;
+      return res;
+    });
+};
+
+type ModuleType = 'cjs' | 'esm';
+const MODULE_TYPES: ModuleType[] = ['cjs', 'esm'];
+const MODULE_TYPE_ESM = MODULE_TYPES[1];
+
+type ExecBuilderParams = { baseDir: string; command: string };
+const EXEC_BUILDER_PARAMS: Record<ModuleType, ExecBuilderParams> = {
+  cjs: {
+    baseDir: 'cjs-path-mapping',
+    command: CMD_TS_NODE_WITHOUT_PROJECT_FLAG,
+  },
+  esm: {
+    baseDir: 'esm-path-mapping',
+    command: CMD_ESM_LOADER_WITHOUT_PROJECT,
+  },
+};
+
+const PROJECT_CONFIGS = {
+  BASE_URL_NO_PATHS: 'tsconfig-baseurl-only.json',
+  BASE_URL_AND_PATHS: 'tsconfig.json',
+  BASE_URL_STAR_PATH: 'tsconfig-star-path.json',
+};
+
+for (const moduleType of MODULE_TYPES) {
+  test.suite(`path mapping ${moduleType}`, (test) => {
+    test.runIf(nodeSupportsEsmHooks || moduleType !== MODULE_TYPE_ESM);
+    const execBuilderParams = EXEC_BUILDER_PARAMS[moduleType];
+
+    for (const project of Object.values(PROJECT_CONFIGS)) {
+      // Create ts-node runner for this config
+      const exec = execBuilder(
+        execBuilderParams.command,
+        execBuilderParams.baseDir,
+        project
+      );
+
+      test.suite(`project: ${project}`, (test) => {
+        test(`fallback to node built-in`, async (t) => {
+          await t.notThrowsAsync(exec('import-node-built-in.ts'));
```

Comment:
The reason I changed this is because when the test fails, it prints the error 3 times!

Reply:
Oh is this because using jest `expect()` has messier output than AVA's built-in assertions?  Yeah, that's annoying.  I've tried to make that better -- I think the improvements have merged in from the `main` branch -- but if you still don't like it, could use `t.is(null)`

Reply:
I've reverted to the suggested style, except that I've used `toBeNull()` instead of `toBe(null)` (because the expect JS docs said it gives better error messages). Currently I have no failing tests, so I guess we'll see!